Graph center

**What is the feature or improvement you would like to see?**

A function that computed the [graph center](https://en.wikipedia.org/wiki/Graph_center), i.e the set of vertices with lowest eccentricity.

**Use cases for the feature**

This is a standard concept in graph theory.

**References**

 - https://en.wikipedia.org/wiki/Graph_center
 - `igraph_eccentricity()`

Thanks for the link. Compiling and running tests worked but when I tried to compile a basic program using the library without using Cmake, I get the following error 

`gcc igraph_test.c -I /usr/local/include/igraph -L /usr/local/lib -ligraph -o igraph_test`

`/usr/bin/ld: /usr/local/lib/libigraph.a(vector.c.o): in function `igraph_vector_floor':
/home/pradeep/repos/igraph/src/core/vector.c:98: undefined reference to `floor'
/usr/bin/ld: /usr/local/lib/libigraph.a(complex.c.o): in function `igraph_complex_polar':
/home/pradeep/repos/igraph/src/math/complex.c:41: undefined reference to `cos'
/usr/bin/ld: /home/pradeep/repos/igraph/src/math/complex.c:42: undefined reference to `sin'
/usr/bin/ld: /usr/local/lib/libigraph.a(complex.c.o): in function `igraph_complex_mod':
/home/pradeep/repos/igraph/src/math/complex.c:59: undefined reference to `hypot'
/usr/bin/ld: /usr/local/lib/libigraph.a(complex.c.o): in function `igraph_complex_arg':
/home/pradeep/repos/igraph/src/math/complex.c:68: undefined reference to `atan2'
/usr/bin/ld: /usr/local/lib/libigraph.a(complex.c.o): in function `igraph_complex_abs':
/home/pradeep/repos/igraph/src/math/complex.c:197: undefined reference to `hypot'
/usr/bin/ld: /usr/local/lib/libigraph.a(complex.c.o): in function `igraph_complex_logabs':
/home/pradeep/repos/igraph/src/math/complex.c:211: undefined reference to `log'
/usr/bin/ld: /usr/local/lib/libigraph.a(complex.c.o): in function `igraph_complex_sqrt':
/home/pradeep/repos/igraph/src/math/complex.c:225: undefined reference to `sqrt'
/usr/bin/ld: /home/pradeep/repos/igraph/src/math/complex.c:225: undefined reference to `sqrt'
/usr/bin/ld: /home/pradeep/repos/igraph/src/math/complex.c:225: undefined reference to `sqrt'
/usr/bin/ld: /home/pradeep/repos/igraph/src/math/complex.c:228: undefined reference to `sqrt'
/usr/bin/ld: /home/pradeep/repos/igraph/src/math/complex.c:228: undefined reference to `sqrt'
/usr/bin/ld: /usr/local/lib/libigraph.a(complex.c.o):/home/pradeep/repos/igraph/src/math/complex.c:228: more undefined references to `sqrt' follow
........`

I am only adding part of the log. Even when I tried to follow the "Compiling with Cmake" it doesn't really generate the binary in `src`. Surprising because it worked before. :thinking:

--- a/doc/structural.xxml
+++ b/doc/structural.xxml
@@ -42,6 +42,7 @@
 <!-- doxrox-include igraph_radius -->
 <!-- doxrox-include igraph_pseudo_diameter -->
 <!-- doxrox-include igraph_vertex_path_from_edge_path -->
+<!-- doxrox-include igraph_graph_center -->
 </section>
 
 <section id="widest-path-related-functions"><title>Widest-path related functions</title>

--- a/doc/structural.xxml
+++ b/doc/structural.xxml
@@ -39,10 +39,10 @@
 <!-- doxrox-include igraph_diameter_dijkstra -->
 <!-- doxrox-include igraph_girth -->
 <!-- doxrox-include igraph_eccentricity -->
+<!-- doxrox-include igraph_graph_center -->
 <!-- doxrox-include igraph_radius -->
 <!-- doxrox-include igraph_pseudo_diameter -->
 <!-- doxrox-include igraph_vertex_path_from_edge_path -->
-<!-- doxrox-include igraph_graph_center -->
 </section>
 
 <section id="widest-path-related-functions"><title>Widest-path related functions</title>

--- a/include/igraph_paths.h
+++ b/include/igraph_paths.h
@@ -182,8 +182,10 @@ IGRAPH_EXPORT igraph_error_t igraph_eccentricity(const igraph_t *graph,
 IGRAPH_EXPORT igraph_error_t igraph_radius(const igraph_t *graph, igraph_real_t *radius,
                                 igraph_neimode_t mode);
 
-IGRAPH_EXPORT igraph_error_t igraph_graph_center(const igraph_t *graph, igraph_vector_t *res,
-                                igraph_neimode_t mode);
+IGRAPH_EXPORT igraph_error_t igraph_graph_center(const igraph_t *graph, 
+                                 igraph_vector_t *res,
+                                 igraph_vs_t vids,
+                                 igraph_neimode_t mode);
 
 IGRAPH_EXPORT igraph_error_t igraph_pseudo_diameter(const igraph_t *graph,
                                          igraph_real_t *diameter,

--- a/include/igraph_paths.h
+++ b/include/igraph_paths.h
@@ -184,7 +184,6 @@ IGRAPH_EXPORT igraph_error_t igraph_radius(const igraph_t *graph, igraph_real_t 
 
 IGRAPH_EXPORT igraph_error_t igraph_graph_center(const igraph_t *graph, 
                                  igraph_vector_t *res,
-                                 igraph_vs_t vids,
                                  igraph_neimode_t mode);
 
 IGRAPH_EXPORT igraph_error_t igraph_pseudo_diameter(const igraph_t *graph,

--- a/include/igraph_paths.h
+++ b/include/igraph_paths.h
@@ -182,6 +182,9 @@ IGRAPH_EXPORT igraph_error_t igraph_eccentricity(const igraph_t *graph,
 IGRAPH_EXPORT igraph_error_t igraph_radius(const igraph_t *graph, igraph_real_t *radius,
                                 igraph_neimode_t mode);
 
+IGRAPH_EXPORT igraph_error_t igraph_graph_center(const igraph_t *graph, igraph_vector_t *res,
+                                igraph_neimode_t mode);
+
 IGRAPH_EXPORT igraph_error_t igraph_pseudo_diameter(const igraph_t *graph,
                                          igraph_real_t *diameter,
                                          igraph_integer_t vid_start,

--- a/src/paths/distances.c
+++ b/src/paths/distances.c
@@ -777,3 +777,50 @@ igraph_error_t igraph_pseudo_diameter_dijkstra(const igraph_t *graph,
 
     return IGRAPH_SUCCESS;
 }
+
+/**
+ * \function igraph_graph_center
+ * \brief Centre vertices of a graph.
+ *
+ * The centre vertices of a graph is calculated by finding the vertices
+ * with the minimum eccentricity
+ *
+ * \param graph The input graph, it can be directed or undirected.
+ * \param res Pointer to an initialized vector, the result is stored
+ *    here.
+ * \param mode What kind of paths to consider for the calculation:
+ *    \c IGRAPH_OUT, paths that follow edge directions;
+ *    \c IGRAPH_IN, paths that follow the opposite directions; and
+ *    \c IGRAPH_ALL, paths that ignore edge directions. This argument
+ *    is ignored for undirected graphs.
+ * \return Error code.
+ *
+ * Time complexity: O(v*(|V|+|E|)), where |V| is the number of
+ * vertices, |E| is the number of edges and v is the number of
+ * vertices for which eccentricity is calculated.
+ *
+ * \sa \ref igraph_eccentricity().
+ *
+ */
+igraph_error_t igraph_graph_center(const igraph_t *graph, igraph_vector_t *res,
+                  igraph_neimode_t mode) {
+
+    igraph_vector_t ecc;
+    igraph_vs_t vids = igraph_vss_all();
+
+    IGRAPH_CHECK(igraph_vector_init(&ecc, 0));
+    IGRAPH_CHECK(igraph_eccentricity(graph, &ecc, vids, mode));
+    if (igraph_vcount(graph)) {
+        igraph_integer_t n = igraph_vector_size(&ecc);
+        igraph_integer_t min_eccentricity = (igraph_integer_t) igraph_vector_min(&ecc);
+
+        for (int i = 0; i < n; i++) {
+            if (VECTOR(ecc)[i] == min_eccentricity) {
+                igraph_vector_push_back(res, i);
+            }
+        }
+    }
+
+    igraph_vector_destroy(&ecc);
+    return 0;
+}

--- a/src/paths/distances.c
+++ b/src/paths/distances.c
@@ -783,8 +783,9 @@ igraph_error_t igraph_pseudo_diameter_dijkstra(const igraph_t *graph,
  * \brief Central vertices of a graph.
  *
  * The central vertices of a graph are calculated by finding the vertices
- * with the minimum eccentricity. The calculation is done per connected component:
- * the central vertices of each component are returned.
+ * with the minimum eccentricity. This concept is typically applied to
+ * connected graphs. In undirected disconnected graphs, the calculation
+ * is effectively done per connected component.
  * 
  * \param graph The input graph, it can be directed or undirected.
  * \param res Pointer to an initialized vector, the result is stored
@@ -796,9 +797,8 @@ igraph_error_t igraph_pseudo_diameter_dijkstra(const igraph_t *graph,
  *    is ignored for undirected graphs.
  * \return Error code.
  *
- * Time complexity: O(v*(|V|+|E|)), where |V| is the number of
- * vertices, |E| is the number of edges and v is the number of
- * vertices for which eccentricity is calculated.
+ * Time complexity: O(|V| (|V|+|E|)), where |V| is the number of
+ * vertices and |E| is the number of edges.
  *
  * \sa \ref igraph_eccentricity().
  *
@@ -810,21 +810,25 @@ igraph_error_t igraph_graph_center(const igraph_t *graph,
     igraph_vector_t ecc;
 
     igraph_vector_clear(res);
+    if (igraph_vcount(igraph) == 0) {
+        return IGRAPH_SUCCESS;
+    }
+    
     IGRAPH_VECTOR_INIT_FINALLY(&ecc, 0);
-    // Assume igraph_eccentricity() does not return infinity or NaN 
     IGRAPH_CHECK(igraph_eccentricity(graph, &ecc, igraph_vss_all(), mode));
-    if (igraph_vcount(graph) > 0) {
-        igraph_real_t n = igraph_vector_size(&ecc);
-        igraph_real_t min_eccentricity = igraph_vector_min(&ecc);
 
-        for (igraph_integer_t i = 0; i < n; i++) {
-            if (VECTOR(ecc)[i] == min_eccentricity) {
-                IGRAPH_CHECK(igraph_vector_push_back(res, i));
-            }
+    /* igraph_eccentricity() does not return infinity or NaN, and the null graph
+     * case was handled above, therefore calling vector_min() is safe. */
+    igraph_real_t min_eccentricity = igraph_vector_min(&ecc);
+    igraph_real_t n = igraph_vector_size(&ecc);
+    for (igraph_integer_t i = 0; i < n; i++) {
+        if (VECTOR(ecc)[i] == min_eccentricity) {
+            IGRAPH_CHECK(igraph_vector_push_back(res, i));
         }
     }
 
     igraph_vector_destroy(&ecc);
     IGRAPH_FINALLY_CLEAN(1);
+
     return IGRAPH_SUCCESS;
 }

--- a/src/paths/distances.c
+++ b/src/paths/distances.c
@@ -805,23 +805,21 @@ igraph_error_t igraph_pseudo_diameter_dijkstra(const igraph_t *graph,
  */
 igraph_error_t igraph_graph_center(const igraph_t *graph, 
                     igraph_vector_t *res,
-                    igraph_vs_t vids,
                     igraph_neimode_t mode) {
 
     igraph_vector_t ecc;
 
     igraph_vector_clear(res);
-    IGRAPH_ASSERT(igraph_vector_empty(res));
     IGRAPH_VECTOR_INIT_FINALLY(&ecc, 0);
     // Assume igraph_eccentricity() does not return infinity or NaN 
-    IGRAPH_CHECK(igraph_eccentricity(graph, &ecc, vids, mode));
+    IGRAPH_CHECK(igraph_eccentricity(graph, &ecc, igraph_vss_all(), mode));
     if (igraph_vcount(graph) > 0) {
         igraph_real_t n = igraph_vector_size(&ecc);
         igraph_real_t min_eccentricity = igraph_vector_min(&ecc);
 
         for (igraph_integer_t i = 0; i < n; i++) {
             if (VECTOR(ecc)[i] == min_eccentricity) {
-                igraph_vector_push_back(res, i);
+                IGRAPH_CHECK(igraph_vector_push_back(res, i));
             }
         }
     }

--- a/src/paths/distances.c
+++ b/src/paths/distances.c
@@ -782,8 +782,9 @@ igraph_error_t igraph_pseudo_diameter_dijkstra(const igraph_t *graph,
  * \function igraph_graph_center
  * \brief Central vertices of a graph.
  *
- * The central vertices of a graph is calculated by finding the vertices
- * with the minimum eccentricity. 
+ * The central vertices of a graph are calculated by finding the vertices
+ * with the minimum eccentricity. The calculation is done per connected component:
+ * the central vertices of each component are returned.
  * 
  * \param graph The input graph, it can be directed or undirected.
  * \param res Pointer to an initialized vector, the result is stored

--- a/src/paths/distances.c
+++ b/src/paths/distances.c
@@ -810,7 +810,7 @@ igraph_error_t igraph_graph_center(const igraph_t *graph,
     igraph_vector_t ecc;
 
     igraph_vector_clear(res);
-    if (igraph_vcount(igraph) == 0) {
+    if (igraph_vcount(graph) == 0) {
         return IGRAPH_SUCCESS;
     }
     

--- a/src/paths/distances.c
+++ b/src/paths/distances.c
@@ -809,6 +809,7 @@ igraph_error_t igraph_graph_center(const igraph_t *graph,
                     igraph_neimode_t mode) {
 
     igraph_vector_t ecc;
+
     igraph_vector_clear(res);
     IGRAPH_ASSERT(igraph_vector_empty(res));
     IGRAPH_VECTOR_INIT_FINALLY(&ecc, 0);

--- a/src/paths/distances.c
+++ b/src/paths/distances.c
@@ -780,11 +780,11 @@ igraph_error_t igraph_pseudo_diameter_dijkstra(const igraph_t *graph,
 
 /**
  * \function igraph_graph_center
- * \brief Centre vertices of a graph.
+ * \brief Central vertices of a graph.
  *
- * The centre vertices of a graph is calculated by finding the vertices
- * with the minimum eccentricity
- *
+ * The central vertices of a graph is calculated by finding the vertices
+ * with the minimum eccentricity. 
+ * 
  * \param graph The input graph, it can be directed or undirected.
  * \param res Pointer to an initialized vector, the result is stored
  *    here.
@@ -802,19 +802,22 @@ igraph_error_t igraph_pseudo_diameter_dijkstra(const igraph_t *graph,
  * \sa \ref igraph_eccentricity().
  *
  */
-igraph_error_t igraph_graph_center(const igraph_t *graph, igraph_vector_t *res,
-                  igraph_neimode_t mode) {
+igraph_error_t igraph_graph_center(const igraph_t *graph, 
+                    igraph_vector_t *res,
+                    igraph_vs_t vids,
+                    igraph_neimode_t mode) {
 
     igraph_vector_t ecc;
-    igraph_vs_t vids = igraph_vss_all();
-
-    IGRAPH_CHECK(igraph_vector_init(&ecc, 0));
+    igraph_vector_clear(res);
+    IGRAPH_ASSERT(igraph_vector_empty(res));
+    IGRAPH_VECTOR_INIT_FINALLY(&ecc, 0);
+    // Assume igraph_eccentricity() does not return infinity or NaN 
     IGRAPH_CHECK(igraph_eccentricity(graph, &ecc, vids, mode));
-    if (igraph_vcount(graph)) {
-        igraph_integer_t n = igraph_vector_size(&ecc);
-        igraph_integer_t min_eccentricity = (igraph_integer_t) igraph_vector_min(&ecc);
+    if (igraph_vcount(graph) > 0) {
+        igraph_real_t n = igraph_vector_size(&ecc);
+        igraph_real_t min_eccentricity = igraph_vector_min(&ecc);
 
-        for (int i = 0; i < n; i++) {
+        for (igraph_integer_t i = 0; i < n; i++) {
             if (VECTOR(ecc)[i] == min_eccentricity) {
                 igraph_vector_push_back(res, i);
             }
@@ -822,5 +825,6 @@ igraph_error_t igraph_graph_center(const igraph_t *graph, igraph_vector_t *res,
     }
 
     igraph_vector_destroy(&ecc);
-    return 0;
+    IGRAPH_FINALLY_CLEAN(1);
+    return IGRAPH_SUCCESS;
 }

--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -294,6 +294,7 @@ add_legacy_tests(
   igraph_get_k_shortest_paths
   igraph_get_shortest_paths2
   igraph_get_shortest_paths_bellman_ford
+  igraph_graph_center
   igraph_is_bipartite
   igraph_is_connected
   igraph_is_chordal

--- a/tests/unit/igraph_graph_center.c
+++ b/tests/unit/igraph_graph_center.c
@@ -1,0 +1,86 @@
+/*
+   IGraph library.
+   Copyright (C) 2021  The igraph development team <igraph@igraph.org>
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+
+#include <igraph.h>
+
+#include "test_utilities.h"
+
+int main() {
+
+    igraph_t g;
+    igraph_vector_t center;
+
+    igraph_vector_init(&center, 0);
+
+    printf("Null graph:\n");
+    igraph_empty(&g, 0, IGRAPH_UNDIRECTED);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
+    igraph_vector_clear(&center);
+    igraph_destroy(&g);
+
+    printf("\nSingleton graph:\n");
+    igraph_empty(&g, 1, IGRAPH_UNDIRECTED);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
+    igraph_vector_clear(&center);
+    igraph_destroy(&g);
+
+    printf("\nPath with isolated vertex:\n");
+    igraph_small(&g, 3, IGRAPH_UNDIRECTED,
+                 0,2,
+                 -1);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
+    igraph_vector_clear(&center);
+    igraph_destroy(&g);
+
+    printf("\nUndirected path graph:\n");
+    igraph_ring(&g, 5, IGRAPH_UNDIRECTED, /* mutual */ 0, /* circular */ 0);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
+    igraph_vector_clear(&center);
+    igraph_destroy(&g);
+
+    printf("\nDirected path graph:\n");
+    igraph_ring(&g, 5, IGRAPH_DIRECTED, /* mutual */ 0, /* circular */ 0);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
+    igraph_vector_clear(&center);
+    igraph_destroy(&g);
+
+    printf("\nUndirected star:\n");
+    igraph_star(&g, 10, IGRAPH_STAR_UNDIRECTED, 0);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
+    igraph_vector_clear(&center);
+    igraph_destroy(&g);
+
+    printf("\nOut-star:\n");
+    igraph_star(&g, 10, IGRAPH_STAR_OUT, 0);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
+    igraph_vector_clear(&center);
+    igraph_destroy(&g);
+
+    printf("\nIn-star:\n");
+    igraph_star(&g, 10, IGRAPH_STAR_OUT, 0);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
+    igraph_vector_clear(&center);
+    igraph_destroy(&g);
+
+    igraph_vector_destroy(&center);
+
+    VERIFY_FINALLY_STACK();
+
+    return 0;
+}

--- a/tests/unit/igraph_graph_center.c
+++ b/tests/unit/igraph_graph_center.c
@@ -1,6 +1,6 @@
 /*
    IGraph library.
-   Copyright (C) 2021  The igraph development team <igraph@igraph.org>
+   Copyright (C) 2022  The igraph development team <igraph@igraph.org>
 
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
@@ -30,13 +30,15 @@ int main() {
 
     printf("Null graph:\n");
     igraph_empty(&g, 0, IGRAPH_UNDIRECTED);
-    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, igraph_vss_all(), IGRAPH_OUT) == IGRAPH_SUCCESS);
+    print_vector(&center);
     igraph_vector_clear(&center);
     igraph_destroy(&g);
 
     printf("\nSingleton graph:\n");
     igraph_empty(&g, 1, IGRAPH_UNDIRECTED);
-    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, igraph_vss_all(), IGRAPH_OUT) == IGRAPH_SUCCESS);
+    print_vector(&center);
     igraph_vector_clear(&center);
     igraph_destroy(&g);
 
@@ -44,37 +46,43 @@ int main() {
     igraph_small(&g, 3, IGRAPH_UNDIRECTED,
                  0,2,
                  -1);
-    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, igraph_vss_all(), IGRAPH_OUT) == IGRAPH_SUCCESS);
+    print_vector(&center);
     igraph_vector_clear(&center);
     igraph_destroy(&g);
 
     printf("\nUndirected path graph:\n");
     igraph_ring(&g, 5, IGRAPH_UNDIRECTED, /* mutual */ 0, /* circular */ 0);
-    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, igraph_vss_all(), IGRAPH_OUT) == IGRAPH_SUCCESS);
+    print_vector(&center);
     igraph_vector_clear(&center);
     igraph_destroy(&g);
 
     printf("\nDirected path graph:\n");
     igraph_ring(&g, 5, IGRAPH_DIRECTED, /* mutual */ 0, /* circular */ 0);
-    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, igraph_vss_all(), IGRAPH_OUT) == IGRAPH_SUCCESS);
+    print_vector(&center);
     igraph_vector_clear(&center);
     igraph_destroy(&g);
 
     printf("\nUndirected star:\n");
     igraph_star(&g, 10, IGRAPH_STAR_UNDIRECTED, 0);
-    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, igraph_vss_all(), IGRAPH_OUT) == IGRAPH_SUCCESS);
+    print_vector(&center);
     igraph_vector_clear(&center);
     igraph_destroy(&g);
 
     printf("\nOut-star:\n");
     igraph_star(&g, 10, IGRAPH_STAR_OUT, 0);
-    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, igraph_vss_all(), IGRAPH_OUT) == IGRAPH_SUCCESS);
+    print_vector(&center);
     igraph_vector_clear(&center);
     igraph_destroy(&g);
 
     printf("\nIn-star:\n");
     igraph_star(&g, 10, IGRAPH_STAR_OUT, 0);
-    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, igraph_vss_all(), IGRAPH_OUT) == IGRAPH_SUCCESS);
+    print_vector(&center);
     igraph_vector_clear(&center);
     igraph_destroy(&g);
 

--- a/tests/unit/igraph_graph_center.c
+++ b/tests/unit/igraph_graph_center.c
@@ -74,7 +74,7 @@ int main() {
 
     printf("\nOut-star:\n");
     igraph_star(&g, 10, IGRAPH_STAR_OUT, 0);
-    IGRAPH_ASSERT(igraph_graph_center(&g, &center, igraph_vss_all(), IGRAPH_OUT) == IGRAPH_SUCCESS);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, igraph_vss_all(), IGRAPH_ALL) == IGRAPH_SUCCESS);
     print_vector(&center);
     igraph_vector_clear(&center);
     igraph_destroy(&g);

--- a/tests/unit/igraph_graph_center.c
+++ b/tests/unit/igraph_graph_center.c
@@ -30,14 +30,14 @@ int main() {
 
     printf("Null graph:\n");
     igraph_empty(&g, 0, IGRAPH_UNDIRECTED);
-    IGRAPH_ASSERT(igraph_graph_center(&g, &center, igraph_vss_all(), IGRAPH_OUT) == IGRAPH_SUCCESS);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
     print_vector(&center);
     igraph_vector_clear(&center);
     igraph_destroy(&g);
 
     printf("\nSingleton graph:\n");
     igraph_empty(&g, 1, IGRAPH_UNDIRECTED);
-    IGRAPH_ASSERT(igraph_graph_center(&g, &center, igraph_vss_all(), IGRAPH_OUT) == IGRAPH_SUCCESS);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
     print_vector(&center);
     igraph_vector_clear(&center);
     igraph_destroy(&g);
@@ -46,42 +46,56 @@ int main() {
     igraph_small(&g, 3, IGRAPH_UNDIRECTED,
                  0,2,
                  -1);
-    IGRAPH_ASSERT(igraph_graph_center(&g, &center, igraph_vss_all(), IGRAPH_OUT) == IGRAPH_SUCCESS);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
+    print_vector(&center);
+    igraph_vector_clear(&center);
+    igraph_destroy(&g);
+
+    printf("\nDisconnected graph:\n");
+    igraph_small(&g, 4, IGRAPH_UNDIRECTED, -1);
+    igraph_graph_center(&g, &center, IGRAPH_OUT);
     print_vector(&center);
     igraph_vector_clear(&center);
     igraph_destroy(&g);
 
     printf("\nUndirected path graph:\n");
     igraph_ring(&g, 5, IGRAPH_UNDIRECTED, /* mutual */ 0, /* circular */ 0);
-    IGRAPH_ASSERT(igraph_graph_center(&g, &center, igraph_vss_all(), IGRAPH_OUT) == IGRAPH_SUCCESS);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
+    print_vector(&center);
+    igraph_vector_clear(&center);
+    igraph_destroy(&g);
+
+    printf("\nUndirected Graph\n");
+    igraph_small(&g, 6, IGRAPH_UNDIRECTED, 0, 1, 1, 2, 0, 2, 3, 0, 4, 1, 2, 5, -1);
+    igraph_graph_center(&g, &center, IGRAPH_OUT);
     print_vector(&center);
     igraph_vector_clear(&center);
     igraph_destroy(&g);
 
     printf("\nDirected path graph:\n");
     igraph_ring(&g, 5, IGRAPH_DIRECTED, /* mutual */ 0, /* circular */ 0);
-    IGRAPH_ASSERT(igraph_graph_center(&g, &center, igraph_vss_all(), IGRAPH_OUT) == IGRAPH_SUCCESS);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
     print_vector(&center);
     igraph_vector_clear(&center);
     igraph_destroy(&g);
 
     printf("\nUndirected star:\n");
     igraph_star(&g, 10, IGRAPH_STAR_UNDIRECTED, 0);
-    IGRAPH_ASSERT(igraph_graph_center(&g, &center, igraph_vss_all(), IGRAPH_OUT) == IGRAPH_SUCCESS);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
     print_vector(&center);
     igraph_vector_clear(&center);
     igraph_destroy(&g);
 
     printf("\nOut-star:\n");
     igraph_star(&g, 10, IGRAPH_STAR_OUT, 0);
-    IGRAPH_ASSERT(igraph_graph_center(&g, &center, igraph_vss_all(), IGRAPH_ALL) == IGRAPH_SUCCESS);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_ALL) == IGRAPH_SUCCESS);
     print_vector(&center);
     igraph_vector_clear(&center);
     igraph_destroy(&g);
 
     printf("\nIn-star:\n");
     igraph_star(&g, 10, IGRAPH_STAR_OUT, 0);
-    IGRAPH_ASSERT(igraph_graph_center(&g, &center, igraph_vss_all(), IGRAPH_OUT) == IGRAPH_SUCCESS);
+    IGRAPH_ASSERT(igraph_graph_center(&g, &center, IGRAPH_OUT) == IGRAPH_SUCCESS);
     print_vector(&center);
     igraph_vector_clear(&center);
     igraph_destroy(&g);

--- a/tests/unit/igraph_graph_center.out
+++ b/tests/unit/igraph_graph_center.out
@@ -1,0 +1,23 @@
+Null graph:
+( )
+
+Singleton graph:
+( 0 )
+
+Path with isolated vertex:
+( 1 )
+
+Undirected path graph:
+( 2 )
+
+Directed path graph:
+( 4 )
+
+Undirected star:
+( 0 )
+
+Out-star:
+( 0 )
+
+In-star:
+( 1 2 3 4 5 6 7 8 9 )

--- a/tests/unit/igraph_graph_center.out
+++ b/tests/unit/igraph_graph_center.out
@@ -7,8 +7,14 @@ Singleton graph:
 Path with isolated vertex:
 ( 1 )
 
+Disconnected graph:
+( 0 1 2 3 )
+
 Undirected path graph:
 ( 2 )
+
+Undirected Graph
+( 0 1 2 )
 
 Directed path graph:
 ( 4 )

